In [2]:
import requests
from bs4 import BeautifulSoup
from lxml import html
import numpy as np
import pandas as pd
from tqdm import tqdm
import time 
import csv
import re

In [126]:
PROXY_ADDRESS = "185.199.229.156"
PROXY_PORT =7492
PROXY_USERNAME = "zhudofyh"
PROXY_PASSWORD = "k1tm10gcjs3u"
# 'Accept-Language' : 'en-US,en;q=0.9',
proxy_credential = {"https": f"http://{PROXY_USERNAME}:{PROXY_PASSWORD}@{PROXY_ADDRESS}:{PROXY_PORT}"}

headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/102.0.0.0 Safari/537.36'}

In [101]:
country_url={'England':'https://www.transfermarkt.com/premier-league/startseite/wettbewerb/GB1',
             'Spain' : 'https://www.transfermarkt.com/laliga/startseite/wettbewerb/ES1',
             'Italy':   'https://www.transfermarkt.com/serie-a/startseite/wettbewerb/IT1',
             'Germany':'https://www.transfermarkt.com/bundesliga/startseite/wettbewerb/L1',
             'France': 'https://www.transfermarkt.com/ligue-1/startseite/wettbewerb/FR1'}

#find All clubs links

In [123]:
links=[]
years=np.arange(2015,2022) 
for country, url in country_url.items():
    for year in years:
        s_url=url +'/plus/?saison_id='+ str(year) 
        r = requests.get(s_url,headers=headers,proxies=proxy_credential) 
        soup = BeautifulSoup(r.content, 'html.parser')

  
        table=soup.find('table', attrs={'class':'items'})
        trs=table.find_all('tr')  
        
        for tr in trs[2:]:
            td=tr.find_all('td')[1].find('a', href=True)['href']
            link='https://www.transfermarkt.com'+ td
            print(link)
            if link not in links:
                links.append(link)

https://www.transfermarkt.com/fc-chelsea/startseite/verein/631/saison_id/2015
https://www.transfermarkt.com/manchester-city/startseite/verein/281/saison_id/2015
https://www.transfermarkt.com/fc-arsenal/startseite/verein/11/saison_id/2015
https://www.transfermarkt.com/manchester-united/startseite/verein/985/saison_id/2015
https://www.transfermarkt.com/fc-liverpool/startseite/verein/31/saison_id/2015
https://www.transfermarkt.com/tottenham-hotspur/startseite/verein/148/saison_id/2015
https://www.transfermarkt.com/fc-everton/startseite/verein/29/saison_id/2015
https://www.transfermarkt.com/fc-southampton/startseite/verein/180/saison_id/2015
https://www.transfermarkt.com/west-ham-united/startseite/verein/379/saison_id/2015
https://www.transfermarkt.com/newcastle-united/startseite/verein/762/saison_id/2015
https://www.transfermarkt.com/leicester-city/startseite/verein/1003/saison_id/2015
https://www.transfermarkt.com/stoke-city/startseite/verein/512/saison_id/2015
https://www.transfermarkt.

In [98]:
headers = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (HTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36',
    'Accept-Language' : 'en-US,en;q=0.9'
}

In [127]:
texts=[]
ids=[] 
names=[] 
seasons=[]
mvs=[]
for link in links:
    print(link) 

    club_page=requests.get(link , headers=headers,proxies=proxy_credential) 
    soup=BeautifulSoup(club_page.content, 'html.parser') 
    tree = html.fromstring(club_page.content)
    even_rows=soup.find('table',attrs={'class':'items'}).find('tbody').find_all('tr',attrs={'class':'even'})
    odd_rows=soup.find('table',attrs={'class':'items'}).find('tbody').find_all('tr',attrs={'class':'odd'})

    for row in even_rows:

        col_1=row.find('td',attrs={'class':'posrela'}) 

        #descriptions
        try:
            text=col_1.find('span').find('a')['title']
           
          
      
        except:
            try:
                text=col_1.find('table').find('tr').find('td',attrs={'class':'hauptlink'}).find('a').find('span')['title']
               
             
            except:
                text='none' 
                
            
        #name of player
        try:
            name=row.find('td',attrs={'class':'posrela'}).find('table').find('td',attrs={'class':'hauptlink'}).find('a')['title'] 
            id=row.find('td',attrs={'class':'posrela'}).find('table').find('td',attrs={'class':'hauptlink'}).find('a')['href'].split('/')[4]
        
        except:
            print('warning') 



        names.append(name)
        ids.append(id) 
        season=link.split('/')[-1] 
        seasons.append(season) 
        if text!=name:
            texts.append(text)
            print(text) 
        else:
            texts.append('none') 

        #market value
        try:
            market_value=row.find('td', attrs={'class':'rechts hauptlink'}).find('a').text
          
        except:
            market_value='-'
        mvs.append(market_value)


        #odd_rows
        for row in odd_rows:

            col_1=row.find('td',attrs={'class':'posrela'})

        try:
            text=col_1.find('span').find('a')['title']
      
        except:
            try:
                text=col_1.find('table').find('tr').find('td',attrs={'class':'hauptlink'}).find('a').find('span')['title']
               
             
            except:
                text='none' 
              
            

        #name of player
        try:
            name=row.find('td',attrs={'class':'posrela'}).find('table').find('td',attrs={'class':'hauptlink'}).find('a')['title'] 
            id=row.find('td',attrs={'class':'posrela'}).find('table').find('td',attrs={'class':'hauptlink'}).find('a')['href'].split('/')[4]
        
        except:
            print('warning') 


        try:
            market_value=row.find('td', attrs={'class':'rechts hauptlink'}).find('a').text
        
        except:
            market_value='-'

        mvs.append(market_value)

        names.append(name)
        ids.append(id) 
        season=link.split('/')[-1] 
        seasons.append(season)
        if text!=name:
            texts.append(text)
             
        else:
            texts.append('none')
  

Streaming output truncated to the last 5000 lines.
https://www.transfermarkt.com/aston-villa/startseite/verein/405/saison_id/2021
On loan from AS Roma; committed until: Jul 1, 2022; fee: €3.50m
Joined as a winter arrival from: Arsenal FC; date: Jan 27, 2022; fee: free transfer
Returned after loan spell with Newcastle United; date: May 31, 2022; fee: End of loan
Joined from Inter Milan; date: Jul 1, 2021; fee: free transfer
Internal winter transfer: Aston Villa U21; Date: Jan 1, 2022
Joined from Bayer 04 Leverkusen; date: Aug 4, 2021; fee: €32.00m
Returned after loan spell with Basaksehir FK; date: Jun 30, 2022; fee: End of loan
Joined from Norwich City; date: Jul 1, 2021; fee: €38.40m
Returned after loan spell with Everton FC; date: Jun 30, 2022; fee: End of loan
Joined from Southampton FC; date: Aug 4, 2021; fee: €29.40m
Returned after loan spell with Preston North End; date: May 31, 2022; fee: End of loan
https://www.transfermarkt.com/fc-everton/startseite/verein/29/saison_id/2021
Jo

In [128]:
len(texts)

25556

In [129]:
len(ids)

25556

In [130]:
table_dict={}
table_dict['player_id']=ids
table_dict['season']=seasons
table_dict['player_name']=names
table_dict['discription']=texts
table_dict['market_value']=mvs
df=pd.DataFrame(data=table_dict)

In [131]:
df.drop_duplicates( keep='last',inplace=True)

In [132]:
df.shape

(13463, 5)

In [133]:
df.isnull().sum()

player_id       0
season          0
player_name     0
discription     0
market_value    0
dtype: int64

In [135]:
df['market_value']=df['market_value'].str.replace('€','')

In [136]:
df.head(20)

player_id season         player_name  \
0      33873   2015       Asmir Begovic   
2      16810   2015        Marco Amelia   
4      27511   2015         Gary Cahill   
6     136610   2015     Papy Djilobodji   
8     157490   2015      Michael Hector   
10    245893   2015         Matt Miazga   
12    316551   2015  Jake Clarke-Salter   
14     57500   2015   César Azpilicueta   
16    236490   2015            Ola Aina   
18     30739   2015      Mikel John Obi   
20     54170   2015             Ramires   
22    236496   2015     Charlie Colkett   
24     59866   2015        Victor Moses   
26    176485   2015        Lucas Piazon   
28     50202   2015         Eden Hazard   
30    281404   2015              Kenedy   
32    131996   2015     Bertrand Traoré   
34     39152   2015      Radamel Falcao   
36     45121   2015           Loïc Rémy   
38    331726   2015       Tammy Abraham   

                                          discription market_value  
0   Joined from Stoke City; date: Jul 13, 2015; fe...       12.00m  
2   Joined from Without Club; date: Oct 8, 2015; f...         250k  
4                                                none       16.00m  
6   Returned after loan spell with SV Werder Breme...        3.50m  
8   Returned after loan spell with Reading FC; dat...        3.00m  
10  Joined as a winter arrival from: New York Red ...        2.50m  
12                                               none            -  
14                                               none       20.00m  
16                                               none            -  
18                                               none       10.00m  
20                                               none       15.00m  
22                                               none            -  
24  Returned after loan spell with West Ham United...        7.50m  
26  Returned after loan spell with Reading FC; dat...        3.00m  
28                                               none       65.00m  
30  Joined from Fluminense Football Club; date: Au...        8.00m  
32  Returned after loan spell with Vitesse Arnhem;...        4.00m  
34  On loan from AS Monaco; started on: Jul 3, 201...       23.00m  
36                                               none       10.00m  
38                                               none            -

In [137]:
df['market_value']=df['market_value'].apply(lambda x: float(x[:-1]) * 1000000 if 'm' in x else float(x[:-2]) * 1000000000 if 'bn' in x else float(x[:-1]) * 1000 if x.endswith('k') else x)

In [141]:
df.head(20)

player_id season         player_name  \
0      33873   2015       Asmir Begovic   
2      16810   2015        Marco Amelia   
4      27511   2015         Gary Cahill   
6     136610   2015     Papy Djilobodji   
8     157490   2015      Michael Hector   
10    245893   2015         Matt Miazga   
12    316551   2015  Jake Clarke-Salter   
14     57500   2015   César Azpilicueta   
16    236490   2015            Ola Aina   
18     30739   2015      Mikel John Obi   
20     54170   2015             Ramires   
22    236496   2015     Charlie Colkett   
24     59866   2015        Victor Moses   
26    176485   2015        Lucas Piazon   
28     50202   2015         Eden Hazard   
30    281404   2015              Kenedy   
32    131996   2015     Bertrand Traoré   
34     39152   2015      Radamel Falcao   
36     45121   2015           Loïc Rémy   
38    331726   2015       Tammy Abraham   

                                          discription market_value  
0   Joined from Stoke City; date: Jul 13, 2015; fe...   12000000.0  
2   Joined from Without Club; date: Oct 8, 2015; f...     250000.0  
4                                                none   16000000.0  
6   Returned after loan spell with SV Werder Breme...    3500000.0  
8   Returned after loan spell with Reading FC; dat...    3000000.0  
10  Joined as a winter arrival from: New York Red ...    2500000.0  
12                                               none            -  
14                                               none   20000000.0  
16                                               none            -  
18                                               none   10000000.0  
20                                               none   15000000.0  
22                                               none            -  
24  Returned after loan spell with West Ham United...    7500000.0  
26  Returned after loan spell with Reading FC; dat...    3000000.0  
28                                               none   65000000.0  
30  Joined from Fluminense Football Club; date: Au...    8000000.0  
32  Returned after loan spell with Vitesse Arnhem;...    4000000.0  
34  On loan from AS Monaco; started on: Jul 3, 201...   23000000.0  
36                                               none   10000000.0  
38                                               none            -

In [143]:
market_value_df=df.drop(['discription','player_name'],axis=1)

In [147]:
index=market_value_df[market_value_df['market_value']=='-'].index.to_list()
market_value_df.drop(index,axis=0,inplace=True)

In [148]:
market_value_df.head()

player_id season market_value
0     33873   2015   12000000.0
2     16810   2015     250000.0
4     27511   2015   16000000.0
6    136610   2015    3500000.0
8    157490   2015    3000000.0

In [149]:
market_value_df.shape

(12594, 3)

In [150]:
market_value_df.to_csv('market_value_df.csv', index=False)

In [233]:
texts=[]
signed_from=[]
ids=[] 
current_club_ids=[]

for link in links:
    url=link.split('startseite/verein')[0]+'kader/verein'+ link.split('verein')[1]+'/plus/1'
    print(url)
    club_page=requests.get(url,headers=headers,proxies=proxy_credential) 
    tree = html.fromstring(club_page.content)
    soup=BeautifulSoup(club_page.content, 'html.parser') 
    even_rows=soup.find('table',attrs={'class':'items'}).find('tbody').find_all('tr',attrs={'class':'even'})
    odd_rows=soup.find('table',attrs={'class':'items'}).find('tbody').find_all('tr',attrs={'class':'odd'})
    for row in even_rows:

        current_club_id=link.split('/')[6] 
        current_club_ids.append(current_club_ids)
        col_1=row.find('td',attrs={'class':'posrela'})

        #descriptions
        try:
            text=col_1.find('span').find('a')['title']
            texts.append(text)
          
      
        except:
            try:
                text=col_1.find('table').find('tr').find('td',attrs={'class':'hauptlink'}).find('a').find('span')['title']
                texts.append(text) 
             
            except:
                text='none' 
                texts.append(text)

        #signed from
        try:
            signed=row.find_all('td',attrs={'class':'zentriert'})[6].find('a')['title'].split(':')[0]
            signed_from.append(signed)
          
        except:
            try:
                 signed=row.find_all('td',attrs={'class':'zentriert'})[7].find('a')['title'].split(':')[0]
                 signed_from.append(signed)
                 
            except:
                 signed='none'
                 signed_from.append(signed)
              


        #name of player
        try:
            name=row.find('td',attrs={'class':'posrela'}).find('table').find('td',attrs={'class':'hauptlink'}).find('a').text
            id=row.find('td',attrs={'class':'posrela'}).find('table').find('td',attrs={'class':'hauptlink'}).find('a')['href'].split('/')[4]
            name=name.split('\n')[1].replace('\xa0','').rstrip().lstrip()

            ids.append(id)
            names.append(name)
        except:
            name=row.find('td',attrs={'class':'posrela'}).find('a').text
            names.append(name)
            ids.append(id)
  
        
        #odd_rows
        for row in odd_rows:
            current_club_id=link.split('/')[6]
            current_club_ids.append(current_club_ids)
            col_1=row.find('td',attrs={'class':'posrela'})
        try:
            text=col_1.find('span').find('a')['title']
            texts.append(text)
          
      
        except:
            try:
                text=col_1.find('table').find('tr').find('td',attrs={'class':'hauptlink'}).find('a').find('span')['title']
                texts.append(text) 
             
            except:
                text='none' 
                texts.append(text)
          
        #signed_from 
        try:
            signed=row.find_all('td',attrs={'class':'zentriert'})[6].find('a')['title'].split(':')[0]
            signed_from.append(signed)
    
        except:
            try:
                 signed=row.find_all('td',attrs={'class':'zentriert'})[7].find('a')['title'].split(':')[0]
                 signed_from.append(signed)
               
            except:
                 signed='none'
                 signed_from.append(signed)
                 

        #name of player
        try:
            name=row.find('td',attrs={'class':'posrela'}).find('table').find('td',attrs={'class':'hauptlink'}).find('a').text
            id=row.find('td',attrs={'class':'posrela'}).find('table').find('td',attrs={'class':'hauptlink'}).find('a')['href'].split('/')[4]
            name=name.split('\n')[1].replace('\xa0','').rstrip().lstrip()
   
            names.append(name)
        except:
            name=row.find('td',attrs={'class':'posrela'}).find('a').text
            names.append(name) 

        ids.append(id)

https://www.transfermarkt.com/fc-chelsea/kader/verein/631/saison_id/2015/plus/1
https://www.transfermarkt.com/manchester-city/kader/verein/281/saison_id/2015/plus/1
https://www.transfermarkt.com/fc-arsenal/kader/verein/11/saison_id/2015/plus/1
https://www.transfermarkt.com/manchester-united/kader/verein/985/saison_id/2015/plus/1
https://www.transfermarkt.com/fc-liverpool/kader/verein/31/saison_id/2015/plus/1
https://www.transfermarkt.com/tottenham-hotspur/kader/verein/148/saison_id/2015/plus/1
https://www.transfermarkt.com/fc-everton/kader/verein/29/saison_id/2015/plus/1
https://www.transfermarkt.com/fc-southampton/kader/verein/180/saison_id/2015/plus/1
https://www.transfermarkt.com/west-ham-united/kader/verein/379/saison_id/2015/plus/1
https://www.transfermarkt.com/newcastle-united/kader/verein/762/saison_id/2015/plus/1
https://www.transfermarkt.com/leicester-city/kader/verein/1003/saison_id/2015/plus/1
https://www.transfermarkt.com/stoke-city/kader/verein/512/saison_id/2015/plus/1
ht

In [236]:
len(ids)

25556

In [248]:
len(texts)

25556

In [249]:
len(seasons)

25556

In [256]:
len(signed_from)

25556

In [599]:
info={}
info['player_id']=ids
info['season']=seasons
info['discription']=texts
description=pd.DataFrame(data=info)

In [600]:
description.drop_duplicates( keep='last',inplace=True)

In [601]:
index=description[description['discription']=='none'].index.to_list()

In [602]:
description.drop(index,axis=0,inplace=True)

In [603]:
description.shape

(2004, 3)

In [604]:
print(len(description['player_id'].unique()))

1081


#we have 1081 unique player id descriptions
include this type of situation:

1. joined from discription is about fee, date and clubs signed_from(transfer) 
2. internal tansfer

3. on loan and Returned after loan

4. injured

5. red card and yellow card

In [605]:
index=description[description['discription'].str.contains('Joined')].index

In [606]:
description['discription'].loc[index]='transfer'

In [607]:
index=description[description['discription'].str.contains('Internal transfer')].index
description['discription'].loc[index]='Internal transfer'

In [608]:
index=description[description['discription'].str.contains('On loan')].index
description['discription'].loc[index]='On loan'

In [609]:
index=description[description['discription'].str.contains('Injury')].index
index

Int64Index([   78,   248,   388,   392,   410,   414,   556,   664,   764,
              968,
            ...
            24542, 24776, 24786, 24816, 25122, 25250, 25284, 25322, 25480,
            25483],
           dtype='int64', length=405)

In [610]:
description['Return_date']=description['discription'].loc[index].apply(lambda x: x.split('on')[1].rstrip().lstrip() if 'Return expected' in x else 'Unknown' if 'Return unknown' in x else '-')

In [611]:
description['discription'].loc[index]='Injury'

In [612]:
index=description[description['discription'].str.contains('Returned after loan')].index
description['Return_date'].loc[index]='-'
description['discription'].loc[index]='Returned'

In [613]:
index=description[description['discription'].str.contains('expected')].index
index

Int64Index([   90,   122,   124,   228,   270,   294,   702,   812,   842,
              920,
            ...
            24412, 24604, 24794, 24818, 24918, 24974, 25398, 25410, 25462,
            25472],
           dtype='int64', length=186)

In [614]:
description['Return_date'].loc[index]=description['discription'].loc[index].apply(lambda x: x.split('on')[-1].rstrip().lstrip() if 'on' in x else x )

In [615]:
description['discription'].loc[index]='Injury'

In [617]:
index=description[description['discription'].str.contains('unknown')].index
index

Int64Index([   34,   252,   528,   537,   610,   616,   896,  1070,  1120,
             1122,
            ...
            23918, 24086, 24274, 24320, 24649, 24664, 24692, 24708, 25046,
            25232],
           dtype='int64', length=278)

In [618]:
description['Return_date'].loc[index]='unknown'

In [619]:
description['discription'].loc[index]='Injury'

In [620]:
index=description[description['discription'].str.contains('Joining')].index
index

Int64Index([  554,   954,  1010,  1020,  1160,  1310,  1504,  1532,  1656,
             1768,
            ...
            23574, 23832, 23966, 24242, 24602, 24968, 25012, 25052, 25378,
            25512],
           dtype='int64', length=199)

In [621]:
description['Return_date'].loc[index]=description['discription'].loc[index].apply(lambda x: x.split('date:')[1].split(';')[0].rstrip().lstrip())

In [622]:
description['discription'].loc[index]='Joining'

In [623]:
index=description[description['discription'].str.contains('suspension')].index

description.loc[index]

player_id season                                        discription  \
320       92571   2015             Red card suspension -  - Europa League   
324      126321   2015              Red card suspension -  - A-League Men   
688       80197   2015  Red card suspension - until Jun 9, 2023 - Copa...   
844       15956   2016  Yellow card suspension - until May 28, 2023 - ...   
1074      92571   2016             Red card suspension -  - Europa League   
...         ...    ...                                                ...   
24942    435485   2021  Indirect card suspension - until May 29, 2023 ...   
24998    485963   2021  Red card suspension - until May 28, 2023 - LaL...   
25078    239743   2021  Yellow card suspension - until May 28, 2023 - ...   
25412     18934   2021  Red card suspension - until May 28, 2023 - cro...   
25488    460245   2021  Yellow card suspension - until May 28, 2023 - ...   

      Return_date  
320           NaN  
324           NaN  
688           NaN  
844           NaN  
1074          NaN  
...           ...  
24942         NaN  
24998         NaN  
25078         NaN  
25412         NaN  
25488         NaN  

[100 rows x 4 columns]

In [624]:
description['Return_date'].loc[index]=description['discription'].loc[index].apply(lambda x: x.split('-')[1].split('until')[1].rstrip().lstrip() if 'until' in x else '-')

In [625]:
description['discription'].loc[index]=description['discription'].loc[index].apply(lambda x: x.split('-')[0].rstrip().lstrip())

In [626]:
index=description[description['discription']=='Internal transfer'].index
description['Return_date'].loc[index]='-'

In [628]:
index=description[description['discription'].str.contains('Suspension')].index
index 

Int64Index([340, 398, 1090, 2007, 5228, 22128, 23576, 25008], dtype='int64')

In [640]:
description['Return_date'].loc[index]=description['discription'].loc[index].apply(lambda x: x.split('-')[1].split('until')[1].rstrip().lstrip() if 'until' in x else '-')

In [641]:
description['discription'].loc[index]=description['discription'].loc[index].apply(lambda x: x.split('-')[0] )

In [650]:
index=description[description['discription']=='transfer'].index
description['Return_date'].loc[index]='-'

In [653]:
index=description[description['discription']=='On loan'].index
description['Return_date'].loc[index]='-'

In [656]:
index=description[description['Return_date'].isnull()].index

In [657]:
description.drop(index,axis=0,inplace=True)

In [658]:
description.isnull().sum()

player_id      0
season         0
discription    0
Return_date    0
dtype: int64

In [659]:
description.shape

(1948, 4)

In [660]:
description.to_csv('description.csv', index=False)